client/asset/btc/txdb.go

```diff
@@ -0,0 +1,408 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package btc
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"decred.org/dcrdex/client/asset"
+	"decred.org/dcrdex/dex"
+	"github.com/dgraph-io/badger"
+)
+
+type extendedWalletTx struct {
+	mtx sync.Mutex
+	*asset.WalletTransaction
+	Confirmed bool `json:"confirmed"`
+	// Create bond transactions are added to the store before
+	// they are submitted.
+	Submitted bool `json:"submitted"`
+}
+
+// "b" and "c" must be the first two prefixes.
+// getPendingTxs relies on this.
+var blockPrefix = []byte("b")
+var pendingPrefix = []byte("c")
+var lastQueryKey = []byte("lq")
+var txPrefix = []byte("t")
+var maxPendingKey = pendingKey(math.MaxUint64)
+
+// pendingKey maps an index to an extendedWalletTransaction. The index is
+// required as there may be multiple pending transactions at the same time.
+func pendingKey(i uint64) []byte {
+	key := make([]byte, len(pendingPrefix)+8)
+	copy(key, pendingPrefix)
+	binary.BigEndian.PutUint64(key[len(pendingPrefix):], i)
+	return key
+}
+
+// blockKey maps a block height and an index to an extendedWalletTransaction.
+// The index is required as there may be multiple transactions in the same
+// block.
+func blockKey(blockHeight, index uint64) []byte {
+	key := make([]byte, len(blockPrefix)+16)
+	copy(key, blockPrefix)
+	binary.BigEndian.PutUint64(key[len(blockPrefix):], blockHeight)
+	binary.BigEndian.PutUint64(key[len(blockPrefix)+8:], index)
+	return key
+}
+
+func parseBlockKey(key []byte) (blockHeight, index uint64) {
+	blockHeight = binary.BigEndian.Uint64(key[len(blockPrefix):])
+	index = binary.BigEndian.Uint64(key[len(blockPrefix)+8:])
+	return
+}
+
+// txKey maps a txid to a blockKey or pendingKey.
+func txKey(txid []byte) []byte {
+	key := make([]byte, len(txPrefix)+len(txid))
+	copy(key, txPrefix)
+	copy(key[len(txPrefix):], txid)
+	return key
+}
+
+type txDB interface {
+	storeTx(tx *extendedWalletTx) error
+	markTxAsSubmitted(txID dex.Bytes) error
+	getTxs(n int, refID *dex.Bytes, past bool) ([]*asset.WalletTransaction, error)
```

Comment:
Please document `getTxs` and other complex areas of code. I see a line like this in `checkPendingTxs`
```Go
_, err = txHistoryDB.getTxs(1, &txID, false)
```
and it raises a bunch of questions. Does the `past` argument matter if `n` is `1`? Is `refID` just a starting point for listing txs, or does that exact key have to be in the db? If you are just checking for the presence of a tx in the db, wouldn't a `getTx` method be appropriate?

I have to look at the implementation for review anyway, but we don't need to doom others to the same fate for eternity. Good docs save man hours.

Reply:
Yeah, it's just getting a single transaction, and you're right, a `getTx` method would be nicer. I'll add some docs.